70_FLIPFluids/main.cpp

```diff
@@ -23,2218 +18,1845 @@ using namespace asset;
+            return false;
+
+        // init grid params
+        usePreset(CENTER_DROP);
+        
+        WorkgroupCountParticles = (numParticles + WorkgroupSize - 1) / WorkgroupSize;
+        WorkgroupCountGrid = {
+            (m_gridData.gridSize.x + WorkgroupGridDim - 1) / WorkgroupGridDim,
+            (m_gridData.gridSize.y + WorkgroupGridDim - 1) / WorkgroupGridDim,
+            (m_gridData.gridSize.z + WorkgroupGridDim - 1) / WorkgroupGridDim
+        };
+
+        {
+            float zNear = 0.1f, zFar = 10000.f;
+            core::vectorSIMDf cameraPosition(14, 8, 12);
+            core::vectorSIMDf cameraTarget(0, 0, 0);
+            matrix4SIMD projectionMatrix = matrix4SIMD::buildProjectionMatrixPerspectiveFovLH(core::radians(60.0f), float(WIN_WIDTH) / WIN_HEIGHT, zNear, zFar);
+            camera = Camera(cameraPosition, cameraTarget, projectionMatrix, 1.069f, 0.4f);
+
+            m_pRenderParams.zNear = zNear;
+            m_pRenderParams.zFar = zFar;
+        }
+        m_pRenderParams.radius = m_gridData.gridCellSize * 0.4f;
+
+        // create buffers
+        video::IGPUBuffer::SCreationParams params = {};
+        params.size = sizeof(SGridData);
+        params.usage = IGPUBuffer::EUF_UNIFORM_BUFFER_BIT | IGPUBuffer::EUF_TRANSFER_DST_BIT | IGPUBuffer::EUF_INLINE_UPDATE_VIA_CMDBUF;
+        createBuffer(gridDataBuffer, params);
+
+        params.size = 2 * sizeof(float32_t4);
+        createBuffer(pressureParamsBuffer, params);
+        
+        params.size = sizeof(SMVPParams);
+        createBuffer(cameraBuffer, params);
+
+        params.size = sizeof(SParticleRenderParams);
+        createBuffer(pParamsBuffer, params);
+
+        params.size = numParticles * sizeof(float32_t3);
+        params.usage = IGPUBuffer::EUF_STORAGE_BUFFER_BIT | IGPUBuffer::EUF_SHADER_DEVICE_ADDRESS_BIT;;
+        createBuffer(particleData.positionBuffer, params, IDeviceMemoryAllocation::EMAF_DEVICE_ADDRESS_BIT);
+        createBuffer(particleData.velocityBuffer, params, IDeviceMemoryAllocation::EMAF_DEVICE_ADDRESS_BIT);
+
+        params.size = numParticles * 6 * sizeof(VertexInfo);
+        createBuffer(particleVertexBuffer, params, IDeviceMemoryAllocation::EMAF_DEVICE_ADDRESS_BIT);
+
+        asset::VkExtent3D gridExtent = { m_gridData.gridSize.x, m_gridData.gridSize.y, m_gridData.gridSize.z };
+
+        // cell materials
+        createGridTexture(gridCellMaterialImageView, asset::EF_R32_UINT, gridExtent, asset::IImage::EUF_STORAGE_BIT, "cell material0");
+        createGridTexture(tempCellMaterialImageView, asset::EF_R32_UINT, gridExtent, asset::IImage::EUF_STORAGE_BIT, "cell material1");
+
+        createGridTexture(gridAxisCellMaterialImageView, asset::EF_R32G32B32A32_UINT, gridExtent, asset::IImage::EUF_STORAGE_BIT, "axis cell material0");
+        createGridTexture(tempAxisCellMaterialImageView, asset::EF_R32G32B32A32_UINT, gridExtent, asset::IImage::EUF_STORAGE_BIT, "axis cell material1");
```

Comment:
why are the axis-cell-material images so "thick" ? (128 bits of uint)

Reply:
they shouldn't even exist, should be entirely done in shared memory, the little compute dispatches for producing this image specifically could disappear